### PR TITLE
[#161387488] Send metrics to Prometheus from paas-metrics

### PR DESCRIPTION
**Pull request by bandesz**
# What

We extended the paas-metrics app to send metrics to Prometheus while still sending the metrics to Datadog.

Changes:
 * Added a /metrics endpoint which exposes the Prometheus metrics for scraping
 * We added a Unit field to the collected metrics, so we can use it as a suffix for the Prom. metrics (e.g. if the Unit is "seconds" and the metric is "foo.bar", then the Prom. metric name will be "paas_foo_bar_seconds")
 * The CF_SKIP_SSL_VALIDATION env var was removed from the app as it's not needed anymore
 * Use Go 1.11

# How to review

1. Deploy your CF pipeline from this branch with DataDog enabled.
  ```
  ENABLE_DATADOG=true BRANCH=paas_metrics_prometheus_161387488 make dev pipelines
  ```

2. Check if the metrics app is serving the metrics:

```
https://paas-metrics.${DEPLOY_ENV}.dev.cloudpipelineapps.digital/metrics
```

3. Check if Prometheus has the new metrics (prefixed with `paas_`)

```
https://prometheus.${DEPLOY_ENV}.dev.cloudpipeline.digital/graph?g0.range_input=1h&g0.expr=paas_aiven_estimated_cost_pounds&g0.tab=0
```

4. Check if these metrics are also available in DataDog for your env.

# Who can review

Not me.

### CODE REVIEW

**Comment by bandesz** [comment]
> Despite the PR comment, I _shall_ Review this :-)

what PR comment? ;)

**Comment by jpluscplusm** [comment]
> what PR comment? ;)

`Who can review` ... :-P

File: tools/metrics/gauges_cf.go [hunk 1]
```diff
@@ -52,30 +52,35 @@ func QuotaGauge(c *Client, interval time.Duration) MetricReadCloser {
...
 				Name:  "op.quota.memory.allocated", // memory allocated to apps
 				Value: float64(allocatedMemory),
+				Unit:  "megabytes",
 			},
 			{
 				Kind:  Gauge,
 				Time:  time.Now(),
 				Name:  "op.quota.routes.reserved", // number of routes reserved
 				Value: float64(reservedRoutes),
+				Unit:  "megabytes",
```
Comment (jpluscplusm) [inline tools/metrics/gauges_cf.go]:
```suggestion
				Unit:  "routes",
```

or 

```suggestion
				Unit:  "count",
```

... though do we want to get into a metric taxonomy discussion? Probably not :-)

File: tools/metrics/main.go [hunk 2]
```diff
@@ -17,54 +21,81 @@ import (
...
+	}
+
+	return 8080
+}
+
+func runHTTPServer(port int, metricsHandler http.Handler) {
+	addr := fmt.Sprintf(":%d", port)
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Cache-Control", "max-age=0,no-store,no-cache")
```
Comment (jpluscplusm) [inline tools/metrics/main.go]: Perhaps include `s-maxage=0` here, too, as we're behind CloudFront.

File: tools/metrics/main.go [hunk 3]
```diff
@@ -86,24 +117,38 @@ func Main() error {
 		}, 30*time.Second),
 		CDNTLSValidityGauge(logger, tlsChecker, cfs, 1*time.Hour),
 		ElasticCacheInstancesGauge(logger, ecs, 5*time.Minute),
+		//TestCounterGauge(10 * time.Second),
```
Comment (jpluscplusm) [inline tools/metrics/main.go]:
```suggestion
```

File: tools/metrics/main.go [hunk 4]
```diff
@@ -86,24 +117,38 @@ func Main() error {
...
+			AppKey:        os.Getenv("DATADOG_APP_KEY"),
+			Logger:        logger,
+			BatchSize:     100,
+			BatchInterval: 60 * time.Second,
+			Tags:          []string{"deploy_env:" + os.Getenv("DEPLOY_ENV")},
+		})
+		multiWriter.AddWriter(dataDogreporter)
+	}
+
+	if os.Getenv("DEBUG") == "1" {
```
Comment (jpluscplusm) [inline tools/metrics/main.go]: Nice.

File: tools/metrics/prometheus_reporter.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+// NewPrometheusReporter creates a new PrometheusReporter instance
+func NewPrometheusReporter(registry prometheus.Registerer) *PrometheusReporter {
+	return &PrometheusReporter{
+		registry:   registry,
+		metricVecs: map[string]interface{}{},
+	}
+}
+
+// WriteMetrics implements MetricWriter. Metrics are buffered
+// in batches not written immediately to rate limit api calls
```
Comment (jpluscplusm) [inline tools/metrics/prometheus_reporter.go]: JOOI whose API are we talking about protecting here?

File: tools/metrics/Gopkg.lock [hunk 6]
```diff
@@ -3,11 +3,14 @@
 
 [[projects]]
   branch = "master"
+  digest = "1:79b84ded24b1338dedccd54b7f67ec966407ce4634bd5ccb2560da762619387e"
```
Comment (jpluscplusm) [inline tools/metrics/Gopkg.lock]: I'm not sure quite what to *suggest*, but these SHA->SHA changes (throughout Gopkg.lock) seem quite arbitrary. Can we embed any more info around them, and what they actually reflect? e.g. tags, branches, dates? As potentially the next-person-to-work-on-this, I don't have great visibility of when I should be modifying these (or allowing tooling to modify them) next ... !

File: tools/metrics/aiven.go [hunk 7]
```diff
@@ -91,6 +90,7 @@ func AivenCostGauge(client *AivenClient, interval time.Duration) MetricReadClose
 					Time:  time.Now(),
 					Name:  "aiven.estimated.cost",
 					Value: currentCost,
+					Unit:  "pounds",
```
Comment (jpluscplusm) [inline tools/metrics/aiven.go]: I've checked the Aiven API docs and can't see a unit, here. Are we sure it's £££ not $$$?

File: tools/metrics/main.go [hunk 8]
```diff
@@ -86,24 +117,38 @@ func Main() error {
...
+		multiWriter.AddWriter(StdOutWriter{})
```
Comment (jpluscplusm) [inline tools/metrics/main.go]: (Nice.)

File: tools/metrics/gauges_cf.go (hunk 1 above)
Comment (bandesz) [reply tools/metrics/gauges_cf.go]: This was a mistake, I'll change it to count.

File: tools/metrics/main.go (hunk 2 above)
Comment (bandesz) [reply tools/metrics/main.go]: Originally we weren't, did this change recently? Also this is a copy-paste from the original code and I don't think it matters as this is checked by gorouter internally as part of the http check and there is no cache in play. (Or I'm badly mistaken)

File: tools/metrics/prometheus_reporter.go (hunk 5 above)
Comment (bandesz) [reply tools/metrics/prometheus_reporter.go]: I don't even know where I was copy pasting this from, I'll remove the unrelated bits.

File: tools/metrics/Gopkg.lock (hunk 6 above)
Comment (bandesz) [reply tools/metrics/Gopkg.lock]: I think these are mostly caused by the newer version of dep and this project wasn't touched in a long time. As the lock file contains internal details we shouldn't care about most of the changes.

Comment (bandesz) [reply tools/metrics/Gopkg.lock]: (Just to make it clear these changes were fully generated by a `dep ensure` command)

File: tools/metrics/aiven.go (hunk 7 above)
Comment (bandesz) [reply tools/metrics/aiven.go]: I've checked it in the Aiven console and the invoices are in pounds.

File: tools/metrics/main.go (hunk 2 above)
Comment (alext) [reply tools/metrics/main.go]: Minor point - this won't be behind CloudFront. CloudFront is only used by cdn-routes it's not in place for *.cloudapps.digital.

File: tools/metrics/gauges_cf.go (hunk 1 above)
Comment (bandesz) [reply tools/metrics/gauges_cf.go]: fixed

File: tools/metrics/main.go (hunk 2 above)
Comment (bandesz) [reply tools/metrics/main.go]: I'll leave this as is

File: tools/metrics/main.go (hunk 3 above)
Comment (bandesz) [reply tools/metrics/main.go]: removed

File: tools/metrics/prometheus_reporter.go (hunk 5 above)
Comment (bandesz) [reply tools/metrics/prometheus_reporter.go]: updated

**Comment by bandesz** [comment]
@jpluscplusm All comments were resolved.

File: tools/metrics/main.go (hunk 2 above)
Comment (jpluscplusm) [reply tools/metrics/main.go]: @alext Thanks - I'd made an assumption we had traffic parity between BYOD and non-BYOD apps.

Comment (bandesz) [reply tools/metrics/main.go]: It's always dangerous to make assumptions, so don't make them :)

File: tools/metrics/prometheus_reporter.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+func NewPrometheusReporter(registry prometheus.Registerer) *PrometheusReporter {
+	return &PrometheusReporter{
+		registry:   registry,
+		metricVecs: map[string]interface{}{},
+	}
+}
+
+// WriteMetrics converts the received metrics into Prometheus metrics.
+// Any metrics are registered in the Prometheus registry the first time they created and we update these metrics later.
+// This was we can manage counters correctly.
```
Comment (jpluscplusm) [inline tools/metrics/prometheus_reporter.go]:
```suggestion
// This way we can manage counters correctly.
```

**Comment by jpluscplusm** [comment]
LGTM:

- Tests pass locally
- pipeline deployment results in data in the `jonmtest` env's Prometheus
- pipeline deployment results in data in the `jonmtest` env's Datadog metrics
- autodelete pipeline works, overnight

Merging.

**Pull request merged**